source/language/openpulse.rst

```diff
@@ -0,0 +1,609 @@
+   }
+
+Manipulating frames based on the state of other frames is also permitted:
+
+.. code-block:: javascript
+
+   // Swap phases between two frames
+   const temp = frame1.phase;
+   frame1.phase = frame2.phase;
+   frame2.phase = temp;
+
+Waveforms
+---------
+
+Waveforms are of type ``waveform`` and can either be:
+
+- An array of complex samples which define the points for the waveform envelope
+- An abstract mathematical function representing a waveform. This will later be
+  materialized into a list of complex samples, either by the compiler or the hardware
+  using the parameters provided to the pulse template.
+
+A value of type ``waveform`` is retrieved by explicitly constructing the complex samples
+or by calling one of the built-in waveform template functions. The latter are initialized by
+assigning a ``waveform`` to the result of a (kernel) function call. Note that each of these
+kernel functions takes a type ``length`` as a first argument, since waveforms need to have a
+definite length. Using the hardware dependent ``dt`` unit is recommended for this length,
+since the compiler may need to down-sample a higher precision waveform to physically realize it.
+
+.. code-block:: javascript
+
+   // arbitrary complex samples
+   waveform arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   kernel gaussian(complex[size] amp, length l, length sigma) -> waveform;
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   kernel sech(complex[size] amp, length l, length sigma) -> waveform;
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // square_width is the width of the square waveform component
+   // sigma is the standard deviation of waveform
+   kernel gaussian_square(complex[size] amp, length l, length square_width, length sigma) -> waveform;
+
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   // beta is the Y correction amplitude, see the DRAG paper
+   kernel drag(complex[size] amp, length l, length sigma, float[size] beta) -> waveform;
+
+   // amp is waveform amplitude
+   // l is the overall length of the waveform
+   kernel constant(complex[size] amp, length l) -> waveform;
+
+   // amp is waveform amplitude
+   // l is the overall length of the waveform
+   // frequency is the frequency of the waveform
+   // phase is the phase of the waveform
+   kernel sine(complex[size] amp, length  l, float[size] frequency, angle[size] phase) -> waveform;
+
+We can manipulate the ``waveform`` types using the following signal processing functions to produce
+new waveforms (this list may be updated as more functionality is required).
+
+.. code-block:: javascript
```

Comment:
One useful function I forgot about keeping around was `scale`. I think this is valuable enough and it's even employed in the "Geometric gate" example.

Reply:
Addressed in 36bafc4.